### ISSUE TITLE
Audit sources of shared state across rustc

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#63643]
Infrastructure for parallel compilation has landed in rustc, but the shared state involved has not been fully documented (in terms of what state exists, invariants, atomicity, lock ordering, etc) or assessed for removal by refactoring.

This issue tracks the PR history for parallelization infrastructure, which is being re-audited to seed an initial list of shared state to assess. The initial output of the audit is [this paper doc](https://paper.dropbox.com/doc/rustc-shared-state-audit--Ai__733hU70wACu6fchZrhFMAg-zsOrKjkMZbBYZJWhgXPke), which will ultimately be turned into a set of individual issues.

- [x] #46709
- [x] #46779
- [x] #47906
- [x] #48586
- [x] #48587
- [x] #48690
- [x] #48691
- [x] #48808
- [x] #48811
- [x] #48904
- [x] #48936
- [x] #49030
- [ ] #49045
- [x] #49349
- [x] #49396
- [x] #49558
- [ ] #49732
- [ ] #49834 - Make SelectionCache and EvaluationCache thread-safe
- [ ] #49882 -  More thread-safety changes
- [ ] #49894 - Rename InternedString to LocalInternedString and introduce a new thread-safe InternedString
- [ ] #50108 - Make GlobalCtxt thread-safe
- [ ] #50699 - Blocking Rayon queries
- [ ] #51383 - Run some stuff in parallel
- [ ] #51487 - Make more passes incremental
- [ ] #56614 - Replace LockCell with atomic types
- [ ] #56946 - Add support for using a jobserver with Rayon
- [ ] #57065 - Optimize try_mark_green and eliminate the lock on dep node colors
- [ ] #57232 - Parallelize and optimize parts of HIR map creation
- [ ] #57308 - Make CompileController thread-safe
- [ ] #58010 - Move privacy checking later in the pipeline and make some passes run in parallel
- [ ] #58019  - Combine all builtin late lints and make lint checking parallel
- [ ] #58679 - Refactor passes and pass execution to be more parallel
- [ ] #59540 -  Use arenas to avoid Lrc in queries #1
- [ ] #59545 - Use arenas to avoid Lrc in queries #2
- [ ] #59804  - Clean up jobserver integration
- [ ] #59809 - Make trait_methods_not_found use a lock

**Comment by mark-i-m** [comment]
@aturon It looks like all of the PRs linked above have been merged?

**Comment by mati865** [comment]
@mark-i-m because this issue is about auditing merged PRs 🙃

**Comment by petrochenkov** [comment]
This issue tracks work that was happening in 2019, and the PR list in the description is also relevant to that time.
If there's a new organized effort in this area, we'll make a new issue with a new list of shared data in the compiler that is relevant to 2026+.
